ws-client/src/transport.rs

```diff
@@ -179,74 +183,118 @@ impl<'a> WsTransportClientBuilder<'a> {
 	) -> Result<(Sender, Receiver), WsHandshakeError> {
-		// Try establish the TCP connection.
-		let tcp_stream = {
-			let socket = TcpStream::connect(sockaddr);
-			let timeout = tokio::time::sleep(self.timeout);
-			futures::pin_mut!(socket, timeout);
-			match future::select(socket, timeout).await {
-				future::Either::Left((socket, _)) => {
-					let socket = socket?;
-					if let Err(err) = socket.set_nodelay(true) {
-						log::warn!("set nodelay failed: {:?}", err);
-					}
-					match tls_connector {
-						None => TlsOrPlain::Plain(socket),
-						Some(connector) => {
-							let dns_name = DNSNameRef::try_from_ascii_str(&self.target.host)?;
-							let tls_stream = connector.connect(dns_name, socket).await?;
-							TlsOrPlain::Tls(tls_stream)
-						}
-					}
-				}
-				future::Either::Right((_, _)) => return Err(WsHandshakeError::Timeout(self.timeout)),
+		let mut sockaddrs = self.target.sockaddrs;
+		let mut path_and_query = PathBuf::from(self.target.path_and_query);
+		let mut host = self.target.host;
+		let mut host_header = self.target.host_header;
+
+		let mut err = None;
+
+		for _ in 0..MAX_REDIRECTIONS_ALLOWED {
+			let sockaddr = match sockaddrs.pop() {
+				Some(addr) => addr,
+				None => return err.unwrap_or(Err(WsHandshakeError::NoAddressFound(host))),
+			};
+
+			let tcp_stream = connect(sockaddr, self.timeout, &host, &tls_connector).await?;
+			let mut client = WsHandshakeClient::new(
+				BufReader::new(BufWriter::new(tcp_stream)),
+				&host_header,
+				path_and_query.to_str().expect("valid UTF-8 checked by Url::parse; qed"),
+			);
+			if let Some(origin) = self.origin_header.as_ref() {
+				client.set_origin(origin);
 			}
-		};
+			// Perform the initial handshake.
+			match client.handshake().await? {
+				ServerResponse::Accepted { .. } => {
+					let mut builder = client.into_builder();
+					builder.set_max_message_size(self.max_request_body_size as usize);
+					let (sender, receiver) = builder.finish();
+					return Ok((Sender { inner: sender }, Receiver { inner: receiver }));
+				}
 
-		log::debug!("Connecting to target: {:?}", self.target);
-		let mut client = WsRawClient::new(
-			BufReader::new(BufWriter::new(tcp_stream)),
-			&self.target.host_header,
-			&self.target.path_and_query,
-		);
-		if let Some(origin) = self.origin_header.as_ref() {
-			client.set_origin(origin);
+				ServerResponse::Rejected { status_code } => {
+					err = Some(Err(WsHandshakeError::Rejected { status_code }));
+				}
+				ServerResponse::Redirect { status_code, location } => {
+					log::trace!("recv redirection: status_code: {}, location: {}", status_code, location);
+					match url::Url::parse(&location) {
+						// redirection with absolute path => need to lookup.
+						Ok(url) => {
+							let target = Target::parse(url)?;
+							sockaddrs = target.sockaddrs;
+							path_and_query = PathBuf::from(target.path_and_query);
+							host = target.host;
+							host_header = target.host_header;
```

Comment:
With the change above this could be just:

```rust
target = Target::parse(url)?;
```

Would also switch `path_and_query` to be `PathBuf` on the `Target` struct so there is no need to cast it here.